REGRESSION(274853@main): [GLib][MSE] Excessive CPU usage from cached web process after closing ksdk.com page, failure to display web content after reopening page, part 2

philn opened the pull request:
#### fc158b61f4bedc9bc1a7dd01689a64c225c813d2
<pre>
REGRESSION(274853@main): [GLib][MSE] Excessive CPU usage from cached web process after closing ksdk.com page, failure to display web content after reopening page, part 2
<a href="https://bugs.webkit.org/show_bug.cgi?id=274802">https://bugs.webkit.org/show_bug.cgi?id=274802</a>

Reviewed by Xabier Rodriguez-Calvar.

The excessive CPU usage was due to mediaPlayerReadyStateChanged() recursing on itself forever after
being triggered due to the MediaSource closing procedure. In the WPE/GTK ports at least, the
isSuspended() condition is always true in this situation. The proposed workaround is to attempt a
fixed amount of recursions.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerReadyStateChanged):
* Source/WebCore/html/HTMLMediaElement.h:

Canonical link: <a href="https://commits.webkit.org/280171@main">https://commits.webkit.org/280171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1161836a5d19aee6ab33b19ba38e265d067a7bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55907 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58903 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6339 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58033 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45017 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4374 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57936 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33122 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48204 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26154 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29906 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5532 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4482 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60492 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5920 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52446 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48273 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51954 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12388 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31060 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32144 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2889,9 +2889,17 @@ void HTMLMediaElement::changeNetworkStateFromLoadingToIdle()
+                mediaPlayerReadyStateChanged();
+            });
+        } else
+            ALWAYS_LOG(LOGIDENTIFIER, "Possible infinite recursion detected, bailing out");
```
I would make this a warning and add a comment at the beginning of this block of code.

mcatanzaro replied:
If you add a warning, we'll surely get bug reports complaining about the warning. We should only print warnings when something unexpected happens. We know this condition is expected and certain to occur regularly.

(I have a feeling there must be a better solution for this bug, but I know very little about multimedia and don't have any suggestions.)

philn replied:
So you're fine with `ALWAYS_LOG`?

And yes, this is a workaround. I'm not entirely satisfied about this either.
I suspect this isn't an issue on mac ports because they run media stuff in GPUP, so with different runtime behaviour, and with working process suspension support.

mcatanzaro replied:
Personally, I would add a FIXME and not print anything (unless you're OK with receiving bug reports when it prints).

philn committed:
REGRESSION(274853@main): [GLib][MSE] Excessive CPU usage from cached web process after closing ksdk.com page, failure to display web content after reopening page, part 2
https://bugs.webkit.org/show_bug.cgi?id=274802

Reviewed by Xabier Rodriguez-Calvar.

The excessive CPU usage was due to mediaPlayerReadyStateChanged() recursing on itself forever after
being triggered due to the MediaSource closing procedure. In the WPE/GTK ports at least, the
isSuspended() condition is always true in this situation. The proposed workaround is to attempt a
fixed amount of recursions.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaPlayerReadyStateChanged):
* Source/WebCore/html/HTMLMediaElement.h:

Canonical link: https://commits.webkit.org/280171@main
